File: Lostify/lib/core/helper/card_validation.dart
class CardValidation {
  static String stripDigits(String value) =>
      value.replaceAll(RegExp(r'\D'), '');

  static String formatVisaInput(String value) {
    final digits = stripDigits(value);
    final limited = digits.length > 16 ? digits.substring(0, 16) : digits;
    final buffer = StringBuffer();
    for (var i = 0; i < limited.length; i++) {
      if (i > 0 && i % 4 == 0) buffer.write('-');
      buffer.write(limited[i]);
    }
    return buffer.toString();
  }

  static String? validateVisa(String value) {
    final digits = stripDigits(value);
    if (digits.length != 16) {
      return 'Visa card number must be 16 digits';
    }
    if (!digits.startsWith('4')) {
      return 'Visa card number must start with 4';
    }
    return null;
  }

  static String? validateNationalCard(String value) {
    final digits = stripDigits(value);
    if (digits.length != 14) {
      return 'National Card number must be exactly 14 digits';
    }
    return null;
  }

  static String? validateForCardType(String value, String cardTypeName) {
    final lower = cardTypeName.toLowerCase();
    if (lower == 'visa') return validateVisa(value);
    if (lower == 'national card') return validateNationalCard(value);
    if (value.isEmpty) return 'Card number is required';
    if (value.length > 30) return 'Card number must be at most 30 characters';
    return null;
  }

  static String normalizeForSubmit(String value, String cardTypeName) {
    final lower = cardTypeName.toLowerCase();
    if (lower == 'visa' || lower == 'national card') {
      return stripDigits(value);
    }
    return value;
  }
}

File: Lostify-Front-End/src/lib/cardValidation.ts
export function stripVisaDigits(value: string): string {
  return value.replace(/\D/g, '');
}

export function formatVisaInput(value: string): string {
  const digits = stripVisaDigits(value).slice(0, 16);
  const parts: string[] = [];
  for (let i = 0; i < digits.length; i += 4) {
    parts.push(digits.slice(i, i + 4));
  }
  return parts.join('-');
}

export function validateVisa(value: string): string | null {
  const digits = stripVisaDigits(value);
  if (digits.length !== 16) {
    return 'Visa card number must be 16 digits';
  }
  if (!digits.startsWith('4')) {
    return 'Visa card number must start with 4';
  }
  return null;
}

export function validateNationalCard(value: string): string | null {
  const digits = value.replace(/\D/g, '');
  if (digits.length !== 14) {
    return 'National Card number must be exactly 14 digits';
  }
  return null;
}

export function isNationalCardType(name: string | undefined): boolean {
  return name?.toLowerCase() === 'national card';
}

export function normalizeCardNumberForSubmit(
  value: string,
  cardTypeName: string | undefined
): string {
  const lower = cardTypeName?.toLowerCase() ?? '';
  if (lower === 'visa' || lower === 'national card') {
    return stripVisaDigits(value);
  }
  return value;
}
